dev/devicelab/lib/framework/ab.dart

```diff
@@ -2,23 +2,102 @@
+const String kAResultsKeyName = 'default engine results';
+const String kBResultsKeyName = 'local engine results';
+
+const String kBenchmarkResultsType = 'A/B summaries';
+const String kBenchmarkABVersion = '1.0';
+
+void _usage(String error) {
+  stderr.writeln(error);
+  stderr.writeln('Usage:\n');
+  stderr.writeln(_argParser.usage);
+  exitCode = 1;
+}
+
+Future<void> main(List<String> rawArgs) async {
+  ArgResults args;
+  try {
+    args = _argParser.parse(rawArgs);
+  } on FormatException catch (error) {
+    _usage('${error.message}\n');
+    return;
+  }
+
+  if (args.rest.length > 1) {
+    _usage('Too many file name arguments');
+    return;
+  }
+  final String filename = args.rest.isEmpty ? 'ABresults.json' : args.rest[0];
+  final ABTest test = ABTest.fromJsonMap(
+      const JsonDecoder().convert(await File(filename).readAsString()) as Map<String, dynamic>
+  );
+  if (args['raw'] as bool) {
+    print(test.rawResults());
+  }
+  if (args['tsv'] as bool) {
+    print(test.printSummary());
+  }
+}
 
 /// Collects data from an A/B test and produces a summary for human evaluation.
 ///
 /// See [printSummary] for more.
 class ABTest {
-  final Map<String, List<double>> _aResults = <String, List<double>>{};
-  final Map<String, List<double>> _bResults = <String, List<double>>{};
+  ABTest(this.localEngine, this.taskName)
+      : runStart = DateTime.now(),
+        _aResults = <String, List<double>>{},
+        _bResults = <String, List<double>>{};
+
+  ABTest.fromJsonMap(Map<String, dynamic> jsonResults)
+      : localEngine = jsonResults[kLocalEngineKeyName] as String,
+        taskName = jsonResults[kTaskNameKeyName] as String,
+        runStart = DateTime.parse(jsonResults[kRunStartKeyName] as String),
+        _runEnd = DateTime.parse(jsonResults[kRunEndKeyName] as String),
+        _aResults = _convertFrom(jsonResults[kAResultsKeyName] as Map<String, dynamic>),
+        _bResults = _convertFrom(jsonResults[kBResultsKeyName] as Map<String, dynamic>);
+
+  final String localEngine;
+  final String taskName;
+  final DateTime runStart;
+  DateTime _runEnd;
+  DateTime get runEnd => _runEnd;
+
+  final Map<String, List<double>> _aResults;
+  final Map<String, List<double>> _bResults;
+
+  static Map<String, List<double>> _convertFrom(dynamic results) {
+    final Map<String, dynamic> resultMap = results as Map<String, dynamic>;
+    return <String, List<double>> {
+      for (String key in resultMap.keys)
+        key: (resultMap[key] as List<dynamic>).cast<double>()
+    };
+  }
 
   /// Adds the result of a single A run of the benchmark.
   ///
   /// The result may contain multiple score keys.
   ///
   /// [result] is expected to be a serialization of [TaskResult].
   void addAResult(Map<String, dynamic> result) {
-    _addResult(result, _aResults);
+    if (_runEnd == null) {
+      _addResult(result, _aResults);
+    } else {
+      throw StateError('Cannot add results to ABTest after it is finalized');
```

Comment:
nit: I like checking for bad state at the beginning of the function. This way you throw early and you don't need to indent the normal logic.

Reply:
Done and...